app/src/src_nonfreenet/org/breezyweather/sources/smg/SmgService.kt

```diff
@@ -0,0 +1,422 @@
+import org.breezyweather.sources.smg.json.SmgForecastResult
+import org.breezyweather.sources.smg.json.SmgUvResult
+import org.breezyweather.sources.smg.json.SmgWarningResult
+import retrofit2.Retrofit
+import java.text.SimpleDateFormat
+import java.util.Calendar
+import java.util.Locale
+import java.util.TimeZone
+import javax.inject.Inject
+import javax.inject.Named
+import kotlin.text.startsWith
+
+class SmgService @Inject constructor(
+    @ApplicationContext context: Context,
+    @Named("JsonClient") client: Retrofit.Builder,
+) : HttpSource(), MainWeatherSource, SecondaryWeatherSource {
+
+    override val id = "smg"
+
+    // Even in its English bulletins, SMG refers to itself as "SMG" rather than
+    // "Macao Meteorological and Geophysical Bureau". Keep the Portuguese name
+    // for use in the source list. The English name can be used in attributions.
+    override val name by lazy {
+        if (context.currentLocale.code.startsWith("zh")) {
+            "地球物理氣象局"
+        } else {
+            "Serviços Meteorológicos e Geofísicos (SMG)"
+        }
+    }
+    override val privacyPolicyUrl by lazy {
+        with(context.currentLocale.code) {
+            when {
+                startsWith("zh") -> "https://www.smg.gov.mo/zh/subpage/21/page/21"
+                startsWith("pt") -> "https://www.smg.gov.mo/pt/subpage/21/page/21"
+                else -> "https://www.smg.gov.mo/en/subpage/21/page/21"
+            }
+        }
+    }
+
+    override val color = Color.rgb(1, 173, 159)
+    override val weatherAttribution by lazy {
+        with(context.currentLocale.code) {
+            when {
+                startsWith("zh") -> "地球物理氣象局"
+                startsWith("pt") -> "Direcção dos Serviços Meteorológicos e Geofísicos"
+                else -> "Macao Meteorological and Geophysical Bureau"
+            }
+        }
+    }
+
+    private val mApi by lazy {
+        client
+            .baseUrl(SMG_BASE_URL)
+            .build()
+            .create(SmgApi::class.java)
+    }
+
+    private val mCmsApi by lazy {
+        client
+            .baseUrl(SMG_CMS_URL)
+            .build()
+            .create(SmgCmsApi::class.java)
+    }
+
+    override val supportedFeaturesInMain = listOf(
+        SecondaryWeatherSourceFeature.FEATURE_CURRENT,
+        SecondaryWeatherSourceFeature.FEATURE_ALERT,
+        SecondaryWeatherSourceFeature.FEATURE_AIR_QUALITY,
+        SecondaryWeatherSourceFeature.FEATURE_NORMALS
+    )
+
+    override fun isFeatureSupportedInMainForLocation(
+        location: Location,
+        feature: SecondaryWeatherSourceFeature?,
+    ): Boolean {
+        return location.countryCode.equals("MO", ignoreCase = true)
+    }
+
+    override fun requestWeather(
+        context: Context,
+        location: Location,
+        ignoreFeatures: List<SecondaryWeatherSourceFeature>,
+    ): Observable<WeatherWrapper> {
+        val lang = when (context.currentLocale.code.substring(0, 2)) {
```

Comment:
Use `with` and `when` and `startsWith` to avoid crash in case it is unable to `substring` due to the code (for some reason) being lower than 2 characters.

Reply:
Fixed